File: examples/rust-mobile-demo/ios/RustMobileDemo/Plugins/AdMob/AdMobBackendImpl.swift
import Foundation
import IstmoRuntime
#if canImport(GoogleMobileAds)
import GoogleMobileAds
import UIKit
#endif

public final class AdMobFactoryImpl: AdMobFactory {
    public init() {}

    public func create(config: AdMobConfig) async throws -> AdMobBackend {
        #if canImport(GoogleMobileAds)

        let cfg = MobileAds.shared.requestConfiguration
        cfg.testDeviceIdentifiers = config.testDeviceIds
        cfg.tagForChildDirectedTreatment = config.childDirectedTreatment
            ? NSNumber(value: true)
            : NSNumber(value: false)

        _ = await withCheckedContinuation { (cont: CheckedContinuation<Bool, Never>) in
            MobileAds.shared.start { _ in cont.resume(returning: true) }
        }
        #endif
        return AdMobBackendImpl()
    }
}

public final class AdMobBackendImpl: AdMobBackend, HandleReleaser {

    private let lock = NSLock()

    private var interstitials: [NativeHandleId: AnyObject] = [:]
    private var rewardeds: [NativeHandleId: AnyObject] = [:]
    private var banners: [NativeHandleId: AnyObject] = [:]

    public init() {}

    public func load_interstitial(ad_unit_id: String) async throws -> NativeHandleId {
        #if canImport(GoogleMobileAds)
        let request = Request()
        do {
            let ad = try await InterstitialAd.load(with: ad_unit_id, request: request)
            let id = IstmoRuntime.shared.allocHandleId(pluginId: AdMobDispatcher.PLUGIN_ID)
            lock.withLock { interstitials[id] = ad }
            return id
        } catch {
            throw Self.mapAdError(error)
        }
        #else
        throw AdError.notInitialized
        #endif
    }

    public func show_interstitial(ad: NativeHandleId) async throws -> InterstitialOutcome {
        #if canImport(GoogleMobileAds)
        let entry = lock.withLock { interstitials[ad] } as? InterstitialAd
        guard let entry = entry else { throw AdError.unknownAd }
        let presenter = await Self.rootViewController()
        guard let presenter = presenter else {
            throw AdError.`internal`("no root view controller for interstitial present")
        }
        let coordinator = InterstitialCoordinator()
        entry.fullScreenContentDelegate = coordinator
        let outcome = await withCheckedContinuation { (cont: CheckedContinuation<InterstitialOutcome, Never>) in
            coordinator.continuation = cont
            DispatchQueue.main.async {
                entry.present(from: presenter)
            }
        }
        lock.withLock { _ = interstitials.removeValue(forKey: ad) }
        IstmoRuntime.shared.forgetHandle(ad)
        return outcome
        #else
        throw AdError.notInitialized
        #endif
    }

    public func load_rewarded(ad_unit_id: String) async throws -> NativeHandleId {
        #if canImport(GoogleMobileAds)
        let request = Request()
        do {
            let ad = try await RewardedAd.load(with: ad_unit_id, request: request)
            let id = IstmoRuntime.shared.allocHandleId(pluginId: AdMobDispatcher.PLUGIN_ID)
            lock.withLock { rewardeds[id] = ad }
            return id
        } catch {
            throw Self.mapAdError(error)
        }
        #else
        throw AdError.notInitialized
        #endif
    }

    public func show_rewarded(ad: NativeHandleId) async throws -> RewardedOutcome {
        #if canImport(GoogleMobileAds)
        let entry = lock.withLock { rewardeds[ad] } as? RewardedAd
        guard let entry = entry else { throw AdError.unknownAd }
        let presenter = await Self.rootViewController()
        guard let presenter = presenter else {
            throw AdError.`internal`("no root view controller for rewarded present")
        }
        let coordinator = RewardedCoordinator()
        entry.fullScreenContentDelegate = coordinator
        let outcome = await withCheckedContinuation { (cont: CheckedContinuation<RewardedOutcome, Never>) in
            coordinator.continuation = cont
            DispatchQueue.main.async {
                entry.present(from: presenter) {
                    let reward = entry.adReward
                    coordinator.granted = true
                    coordinator.rewardType = reward.type
                    coordinator.rewardAmount = UInt32(truncating: reward.amount)
                }
            }
        }
        lock.withLock { _ = rewardeds.removeValue(forKey: ad) }
        IstmoRuntime.shared.forgetHandle(ad)
        return outcome
        #else
        throw AdError.notInitialized
        #endif
    }

    public func show_banner(request: BannerRequest) async throws -> NativeHandleId {
        #if canImport(GoogleMobileAds)
        let id = IstmoRuntime.shared.allocHandleId(pluginId: AdMobDispatcher.PLUGIN_ID)
        return try await MainActor.run { [weak self] in
            guard let self = self, let root = Self.rootViewControllerNow() else {
                throw AdError.`internal`("no root view controller for banner attach")
            }
            let scale = UIScreen.main.scale
            let frame = CGRect(
                x: CGFloat(request.rect.x) / scale,
                y: CGFloat(request.rect.y) / scale,
                width: CGFloat(request.rect.width) / scale,
                height: CGFloat(request.rect.height) / scale,
            )
            let banner = BannerView(adSize: AdSizeFromCGSize(frame.size))
            banner.adUnitID = request.adUnitId
            banner.rootViewController = root
            banner.frame = frame
            banner.load(Request())
            root.view.addSubview(banner)
            self.lock.withLock { self.banners[id] = banner }
            return id
        }
        #else
        throw AdError.notInitialized
        #endif
    }

    public func update_banner(banner: NativeHandleId, rect: BannerRect) async throws {
        #if canImport(GoogleMobileAds)
        let view = lock.withLock { banners[banner] } as? BannerView
        guard let view = view else { return }
        await MainActor.run {
            let scale = UIScreen.main.scale
            view.frame = CGRect(
                x: CGFloat(rect.x) / scale,
                y: CGFloat(rect.y) / scale,
                width: CGFloat(rect.width) / scale,
                height: CGFloat(rect.height) / scale,
            )
        }
        #endif
    }

    public func hide_banner(banner: NativeHandleId) async throws {
        #if canImport(GoogleMobileAds)
        let view = lock.withLock { banners.removeValue(forKey: banner) } as? BannerView
        IstmoRuntime.shared.forgetHandle(banner)
        guard let view = view else { return }
        await MainActor.run { view.removeFromSuperview() }
        #endif
    }

    public func releaseNativeHandle(_ handleId: UInt64) {

        let interstitial = lock.withLock { interstitials.removeValue(forKey: handleId) }
        let rewarded = lock.withLock { rewardeds.removeValue(forKey: handleId) }
        let banner = lock.withLock { banners.removeValue(forKey: handleId) }
        #if canImport(GoogleMobileAds)
        if let bannerView = banner as? BannerView {
            DispatchQueue.main.async { bannerView.removeFromSuperview() }
        }
        _ = interstitial
        _ = rewarded
        #else
        _ = interstitial
        _ = rewarded
        _ = banner
        #endif
    }

    #if canImport(GoogleMobileAds)
    private static func mapAdError(_ error: Error) -> AdError {
        let nserror = error as NSError

        if nserror.domain == "com.google.admob" || nserror.domain.contains("MobileAdsSDK") {
            switch nserror.code {
            case 1: return .invalidRequest(nserror.localizedDescription)
            case 2: return .network(nserror.localizedDescription)
            case 3: return .noFill
            default: return .`internal`(nserror.localizedDescription)
            }
        }
        if nserror.domain == NSURLErrorDomain {
            return .network(nserror.localizedDescription)
        }
        return .`internal`(nserror.localizedDescription)
    }

    @MainActor
    private static func rootViewControllerNow() -> UIViewController? {
        for scene in UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }) {
            for window in scene.windows where window.isKeyWindow {
                return window.rootViewController
            }
        }
        return nil
    }

    private static func rootViewController() async -> UIViewController? {
        await MainActor.run { rootViewControllerNow() }
    }
    #endif
}

#if canImport(GoogleMobileAds)

private final class InterstitialCoordinator: NSObject, FullScreenContentDelegate {

    var continuation: CheckedContinuation<InterstitialOutcome, Never>?

    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        continuation?.resume(returning: .failedToShow)
        continuation = nil
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        continuation?.resume(returning: .dismissed)
        continuation = nil
    }
}

private final class RewardedCoordinator: NSObject, FullScreenContentDelegate {

    var continuation: CheckedContinuation<RewardedOutcome, Never>?
    var granted: Bool = false
    var rewardType: String = ""
    var rewardAmount: UInt32 = 0

    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        continuation?.resume(returning: RewardedOutcome(
            granted: false, rewardType: "", rewardAmount: 0,
        ))
        continuation = nil
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        continuation?.resume(returning: RewardedOutcome(
            granted: granted, rewardType: rewardType, rewardAmount: rewardAmount,
        ))
        continuation = nil
    }
}

#endif

private extension NSLock {
    @discardableResult
    func withLock<T>(_ body: () -> T) -> T {
        lock()
        defer { unlock() }
        return body()
    }
}

File: examples/rust-mobile-demo/android/app/src/main/java/dev/istmo/runtime/AdMobBackendImpl.kt
package dev.istmo.runtime

import android.app.Activity
import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.PopupWindow
import com.google.android.gms.ads.AdError as GmsAdError
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class AdMobBackendImpl(private val activity: Activity) : AdMobBackend, HandleReleaser {

    companion object {
        private const val TAG = "istmo.admob"
    }

    private val interstitials = ConcurrentHashMap<NativeHandleId, InterstitialAd>()
    private val rewardeds = ConcurrentHashMap<NativeHandleId, RewardedAd>()

    private data class BannerEntry(val view: AdView, val popup: PopupWindow)
    private val banners = ConcurrentHashMap<NativeHandleId, BannerEntry>()

    override fun releaseNativeHandle(handleId: Long) {
        interstitials.remove(handleId)
        rewardeds.remove(handleId)
        banners.remove(handleId)?.let { entry ->
            activity.runOnUiThread { tearDownBanner(entry) }
        }
    }

    override suspend fun load_interstitial(ad_unit_id: String): NativeHandleId {
        val ad: InterstitialAd = try {
            withContext(Dispatchers.Main) {
                suspendCoroutine { cont ->
                    InterstitialAd.load(
                        activity,
                        ad_unit_id,
                        AdRequest.Builder().build(),
                        object : InterstitialAdLoadCallback() {
                            override fun onAdLoaded(ad: InterstitialAd) {
                                cont.resume(ad)
                            }

                            override fun onAdFailedToLoad(err: LoadAdError) {
                                cont.resumeWithException(BackendException(mapLoadError(err)))
                            }
                        },
                    )
                }
            }
        } catch (e: BackendException) {
            throw e
        }
        val handleId = IstmoRuntime.allocHandleId(AdMobDispatcher.PLUGIN_ID)
        interstitials[handleId] = ad
        return handleId
    }

    override suspend fun show_interstitial(ad: NativeHandleId): InterstitialOutcome {
        val entry = interstitials.remove(ad) ?: run {
            IstmoRuntime.forgetHandle(ad)
            throw BackendException(AdError.UnknownAd)
        }
        IstmoRuntime.forgetHandle(ad)
        return withContext(Dispatchers.Main) {
            suspendCoroutine<InterstitialOutcome> { cont ->
                entry.fullScreenContentCallback = object : FullScreenContentCallback() {
                    override fun onAdDismissedFullScreenContent() {
                        cont.resume(InterstitialOutcome.Dismissed)
                    }

                    override fun onAdFailedToShowFullScreenContent(err: GmsAdError) {
                        Log.w(TAG, "interstitial failed to show: ${err.code} ${err.message}")
                        cont.resume(InterstitialOutcome.FailedToShow)
                    }
                }
                entry.show(activity)
            }
        }
    }

    override suspend fun load_rewarded(ad_unit_id: String): NativeHandleId {
        val ad: RewardedAd = withContext(Dispatchers.Main) {
            suspendCoroutine { cont ->
                RewardedAd.load(
                    activity,
                    ad_unit_id,
                    AdRequest.Builder().build(),
                    object : RewardedAdLoadCallback() {
                        override fun onAdLoaded(ad: RewardedAd) {
                            cont.resume(ad)
                        }

                        override fun onAdFailedToLoad(err: LoadAdError) {
                            cont.resumeWithException(BackendException(mapLoadError(err)))
                        }
                    },
                )
            }
        }
        val handleId = IstmoRuntime.allocHandleId(AdMobDispatcher.PLUGIN_ID)
        rewardeds[handleId] = ad
        return handleId
    }

    override suspend fun show_rewarded(ad: NativeHandleId): RewardedOutcome {
        val entry = rewardeds.remove(ad) ?: run {
            IstmoRuntime.forgetHandle(ad)
            throw BackendException(AdError.UnknownAd)
        }
        IstmoRuntime.forgetHandle(ad)
        return withContext(Dispatchers.Main) {
            suspendCoroutine<RewardedOutcome> { cont ->
                var granted: RewardedOutcome? = null
                entry.fullScreenContentCallback = object : FullScreenContentCallback() {
                    override fun onAdDismissedFullScreenContent() {
                        cont.resume(granted ?: RewardedOutcome(false, "", 0u))
                    }

                    override fun onAdFailedToShowFullScreenContent(err: GmsAdError) {
                        Log.w(TAG, "rewarded failed to show: ${err.code} ${err.message}")
                        cont.resume(RewardedOutcome(false, "", 0u))
                    }
                }
                entry.show(activity) { reward ->
                    granted = RewardedOutcome(
                        granted = true,
                        rewardType = reward.type,
                        rewardAmount = reward.amount.toUInt(),
                    )
                }
            }
        }
    }

    override suspend fun show_banner(request: BannerRequest): NativeHandleId {
        val handleId = IstmoRuntime.allocHandleId(AdMobDispatcher.PLUGIN_ID)
        withContext(Dispatchers.Main) {
            val view = AdView(activity)
            view.adUnitId = request.adUnitId
            val adSize = pickAdSize(request.rect)
            view.setAdSize(adSize)
            val paintedSizePx = adSizePixels(adSize)
            Log.i(
                TAG,
                "banner request unit=${request.adUnitId} rectPx=${request.rect.width}x${request.rect.height}@${request.rect.x},${request.rect.y} " +
                    "-> adSize=${adSize.width}x${adSize.height}dp = ${paintedSizePx.width}x${paintedSizePx.height}px",
            )
            view.adListener = object : AdListener() {
                override fun onAdLoaded() {
                    Log.i(TAG, "banner loaded (${request.adUnitId})")
                }

                override fun onAdFailedToLoad(err: LoadAdError) {
                    Log.w(
                        TAG,
                        "banner failed to load: code=${err.code} domain=${err.domain} msg=${err.message} " +
                            "cause=${err.cause?.message}",
                    )
                }

                override fun onAdImpression() {
                    Log.i(TAG, "banner impression recorded")
                }
            }
            val popup = attachBannerAsPopup(view, request.rect, paintedSizePx)
            view.loadAd(AdRequest.Builder().build())
            banners[handleId] = BannerEntry(view, popup)
        }
        return handleId
    }

    override suspend fun update_banner(banner: NativeHandleId, rect: BannerRect) {
        val entry = banners[banner] ?: return
        withContext(Dispatchers.Main) {
            val painted = adSizePixels(entry.view.adSize ?: AdSize.BANNER)
            entry.popup.update(rect.x.toInt(), rect.y.toInt(), painted.width, painted.height)
        }
    }

    override suspend fun hide_banner(banner: NativeHandleId) {
        val entry = banners.remove(banner) ?: run {
            IstmoRuntime.forgetHandle(banner)
            return
        }
        IstmoRuntime.forgetHandle(banner)
        withContext(Dispatchers.Main) { tearDownBanner(entry) }
    }

    private data class PxSize(val width: Int, val height: Int)

    private fun attachBannerAsPopup(
        view: AdView,
        rect: BannerRect,
        painted: PxSize,
    ): PopupWindow {
        val host = FrameLayout(activity)
        host.addView(view, FrameLayout.LayoutParams(painted.width, painted.height))
        val popup = PopupWindow(host, painted.width, painted.height, false).apply {
            isClippingEnabled = false
            inputMethodMode = PopupWindow.INPUT_METHOD_NOT_NEEDED
            isTouchable = true
            isFocusable = false
        }
        val decor = activity.window.decorView
        if (decor.windowToken != null) {
            popup.showAtLocation(decor, Gravity.TOP or Gravity.START, rect.x.toInt(), rect.y.toInt())
        } else {
            decor.post {
                popup.showAtLocation(decor, Gravity.TOP or Gravity.START, rect.x.toInt(), rect.y.toInt())
            }
        }
        return popup
    }

    private fun tearDownBanner(entry: BannerEntry) {
        try {
            entry.popup.dismiss()
        } catch (t: Throwable) {
            Log.w(TAG, "banner popup dismiss threw: ${t.message}")
        }
        (entry.view.parent as? ViewGroup)?.removeView(entry.view)
        entry.view.destroy()
    }

    private fun pickAdSize(rect: BannerRect): AdSize {
        val density = activity.resources.displayMetrics.density
        val widthDp = (rect.width.toFloat() / density).toInt().coerceAtLeast(0)
        if (widthDp < AdSize.BANNER.width) {
            return AdSize.BANNER
        }
        return AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(activity, widthDp)
    }

    private fun adSizePixels(size: AdSize): PxSize {
        val widthPx = if (size.width > 0) size.getWidthInPixels(activity) else 0
        val heightPx = if (size.height > 0) size.getHeightInPixels(activity) else 0
        return PxSize(widthPx, heightPx)
    }

    private fun mapLoadError(err: LoadAdError): AdError = when (err.code) {
        3 -> AdError.NoFill
        2 -> AdError.Network(err.message)
        1 -> AdError.InvalidRequest(err.message)
        else -> AdError.Internal("code=${err.code} ${err.message}")
    }
}

